Add support for Sensirion SPS30 Particulate Matter sensors

valordk opened the pull request:
## Description:
Added support for Sensirion SPS30 Particulate Matter sensors

**Related issue (if applicable):** fixes https://github.com/esphome/feature-requests/issues/442

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#442

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Commit:
Add support for Sensirion SPS30 Particulate Matter sensors

OttoWinter commented on esphome/components/sps30/sps30.cpp:
```diff
@@ -0,0 +1,256 @@
+void SPS30Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up sps30...");
+  this->write_command_(SPS30_CMD_SOFT_RESET);
+  delayMicroseconds(500000);
```
```suggestion
  delay(500);
```

1. This is a very excessive delay, using `delay()` here will at least allow some background activity to occur (and the WDT won't go crazy).
2. Is this long timeout really needed? Does the device take that long to boot up?
3. If it is indeed half a second, then this needs to be handled differently - stalling the main thread for that long is not good. You can call SOFT_RESET from setup() and use `set_timeout` after that to delay the rest of the setup with an asynchronous timeout.

valordk replied:
Usually it does not, but in rare situations where the ESP32 restarts without repowering the sensor (i.g. when flashing new firmware) the initialization fails and the component gets marked as failed.
I will wrap the deferred logic into a an async block.

OttoWinter commented on esphome/components/sps30/sps30.cpp:
```diff
@@ -0,0 +1,256 @@
+    this->mark_failed();
+    return;
+  }
+  ESP_LOGD(TAG, "  Firmware version v%0d.%02d", (uint16_t(raw_firmware_version[0]) >> 8),
```
```suggestion
  ESP_LOGD(TAG, "  Firmware version v%0d.%02d", (raw_firmware_version[0] >> 8),
```

The value is already an `uint16_t`

valordk replied:
Fixed :)

OttoWinter commented on esphome/components/sps30/sps30.cpp:
```diff
@@ -0,0 +1,256 @@
+  }
+
+  for (size_t i = 0; i < 8; ++i) {
+    this->serial_number_[i * 2] = uint16_t(uint16_t(raw_serial_number[i]) >> 8);
```
```suggestion
    this->serial_number_[i * 2] = static_cast<char>(raw_serial_number[i]) >> 8);
```

The type of serial_number_ is not uint16_t and raw_serial_number already is an uint16_t

valordk replied:
Fixed

OttoWinter commented on esphome/components/sps30/sps30.cpp:
```diff
@@ -0,0 +1,256 @@
+    this->serial_number_[i * 2] = uint16_t(uint16_t(raw_serial_number[i]) >> 8);
+    this->serial_number_[i * 2 + 1] = uint16_t(uint16_t(raw_serial_number[i] & 0xFF));
+  }
+  ESP_LOGD(TAG, "  Serial Number: '%s'", this->serial_number_);
```
- The serial number is transmitted in ASCII form, right?
- We should make sure the string is null-terminated, add a `this->serial_number_[16] = '\0';` line above this one.

valordk replied:
Yea, otherwise it would not work. The serial number has fixed length.The null termination is added during the declaration, in sps30.h file. Let know if you think it's bad practice.
`char serial_number_[17] = {0};  /// Terminating NULL character`
Setting the serial_number_[16] as null 'corrupts' the serial number

OttoWinter commented on esphome/components/sps30/sps30.cpp:
```diff
@@ -0,0 +1,256 @@
+    ESP_LOGD(TAG, "Retrying to reconnect the sensor.");
+    if (this->write_command_(SPS30_CMD_SOFT_RESET)) {
+      /// Sensor restarted and reading attempt made next cycle
+      delayMicroseconds(500000);
```
Same here, stalling the main thread this long is not good.

valordk replied:
Thanks, I see that. I will wrap this into a set_timeout.

OttoWinter commented on esphome/components/sps30/sps30.cpp:
```diff
@@ -0,0 +1,256 @@
+      this->status_clear_warning();
+      ESP_LOGD(TAG, "Sensor reconnected!");
+    } else {
+      ESP_LOGD(TAG, "Sensor offline?");
```
```suggestion
      ESP_LOGD(TAG, "Sensor offline?");
```

This needs a longer message - something like `Reconnect Failed! Is the sensor offline?`

valordk replied:
I've refactored the debug messages in this block

> [13:15:49][W][i2c:070]: Received NACK on transmit of address 0x69
> [13:15:49][D][sps30:112]: Sensor soft-reset failed. Is the sensor offline?
> [13:15:59][D][sps30:104]: Retrying to reconnect the sensor.
> [13:15:59][D][sps30:110]: Sensor reconnected. Resuming continuous measurements!

OttoWinter commented on esphome/components/sps30/sps30.cpp:
```diff
@@ -0,0 +1,256 @@
+    } else {
+      ESP_LOGD(TAG, "Sensor offline?");
+    }
+  } else {
```
```suggestion
    return;
  }
```

Add the return block there so we can save on indentation for the main block of this function.

valordk replied:
Thanks. I've refactored it.

OttoWinter commented on esphome/components/sps30/sps30.cpp:
```diff
@@ -0,0 +1,256 @@
+
+    uint16_t raw_read_status[1];
+    if (!this->read_data_(raw_read_status, 1) || raw_read_status[0] == 0x00) {
+      ESP_LOGW(TAG, "Data not ready yet!");
```
How often does this happen? If it happens regulary, I think it should be debug/verbose level.

valordk replied:
Not often. I've changed it to Debug level and added clarity to the message

> Sensor measurement not ready yet.

OttoWinter commented on tests/test1.yaml:
```diff
@@ -593,6 +593,45 @@ sensor:
+    pm_1_0:
+      name: "Workshop PM <1µm Weight concentration"
+      id: "workshop_PM_1_0"
+      accuracy_decimals: 2
```
Is there a reason for these `accuracy_decimals: 2`? If the sensor provides more accurate data, you can change the default accuracy in the `sensor.sensor_schema` python line.

valordk replied:
I owerlooked the fact that this could be set on the sensor_schema. I've moved the defaults into the python line. I'll update the test and the documentation to reflect that.

Commit:
Remove blocking of the main thread on initialization;
Improve wording on the debug messages;
Add robustness in re-initialization of reconnected or replaced sensors;

valordk commented:
I've realized that, when quickly disconnected and reconnected (even before the esphome marks the sensor as failed), the flow ends up in a branch where it skips reading the measurement due to sensor not being ready, indefinitely. To address that, I've implemented a maximum amount of skips (5) so that if it happens the sensor is forced to be reset and reconnected for continuous measurement again.

Commit:
Fix code formatting;

OttoWinter commented on esphome/components/sps30/sps30.cpp:
```diff
@@ -0,0 +1,268 @@
+    }
+    return;
+  }
+  /// Check if measurement is ready before reading the value
```
Just FYI, this line will run before setup is completed fully.

The update() function starts running after setup() terminates (which happens immediately, without the 500ms timeout due to the set_timeout).

Could this be an issue or does the code currently handle that fine?

valordk replied:
The code handles it fine. Since the measurement won't be available during the first cycle most of the block will be skipped until the next cycle.
